### PR TITLE
[Workflow] Draft: Add Enum Support to Workflow component

### LINKED ISSUE #44211: Support Enum as workflow places
### Discussed in https://github.com/symfony/symfony/discussions/44203

<div type='discussions-op-text'>

<sup>Originally posted by **stephanvierkant** November 22, 2021</sup>
I've tried the new Enum in PHP 8.1 and I love that Symfony supports this new feature in the Form component.

I'd love to have this support in the Workflow component as well. 

I've changed my status attribute from a string to an Enum field and now this doesn't work anymore:

`$workflow->can($entity, $transition->getName()`

I'd be great is Symfony would support Enum in the Workflow Component as well! </div>

### PR DESCRIPTION
**Pull request by bartrail**
| Q             | A
| ------------- | ---
| Branch?       | 6.3 
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no
| Tickets       | Fix #44211
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/Workflow/Definition.php [hunk 1]
```diff
@@ -48,15 +48,15 @@ public function __construct(array $places, array $transitions, string|array $ini
     }
 
     /**
-     * @return string[]
+     * @return Place[]
```
Comment (derrabus) [inline src/Symfony/Component/Workflow/Definition.php:51]: Switching from strings to your new class everywhere will break a lot of applications that currently use the component. You'll need to find another way.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by OskarStark** [comment]
This was rejected in the past. I don't have the PR/issue at hand

**Comment by derrabus** [comment]
> This was rejected in the past. I don't have the PR/issue at hand

#44211, the issue is linked in the PR description. 🙂 

The issue was closed after @lyrixx' assessment that implementing the feature is not worth the effort. A good PR could prove the opposite.

**Comment by OskarStark** [comment]
Sorry, I overlooked the referenced issue. 

> A good PR could prove the opposite.

Sure, go ahead then 👍

**Comment by stof** [comment]
This PR contains lots of BC breaks, and so it is a no-go as is. That's one of the reason why @lyrixx assessed that the feature is probably not worth the effort: you did less than half of the effort for now.
Side note: using `UnitEnum::$name` as fallback for `BackedEnum::$value` when an enum is not backed does not make sense. Names and values are not the same concept (and this probably indicates a bad integration of enums in the component)

**Comment by bartrail** [comment]
> This PR contains lots of BC breaks, and so it is a no-go as is. That's one of the reason why @lyrixx assessed that the feature is probably not worth the effort: you did less than half of the effort for now. Side note: using `UnitEnum::$name` as fallback for `BackedEnum::$value` when an enum is not backed does not make sense. Names and values are not the same concept (and this probably indicates a bad integration of enums in the component)

I just opened the PR as _draft_ so nothing is finished for now :) I am sure aware of the BC breaks and most of the tests are failing. So I will have a closer look at the BC concerns.

Regarding the `UnitEnum` vs `BackedEnum` case, I have started with `UnitEnum` as the least common denominator, but realised quick that it requires some custom code in the used Enum which feels a bit clunky. I would suggest using `BackedEnum` because it's more feasible.

**Comment by lyrixx** [comment]
Hello, Thanks for the effort, But I'm going to close this PR for the reason explained.

But don't hesitate to re-open when finish / ready.

**Pull request closed without merging**